cmd/podman/system/connection/add.go

```diff
@@ -202,30 +204,69 @@ func getUserInfo(uri *url.URL) (*url.Userinfo, error) {
 	return url.User(usr.Username), nil
 }
 
-func getUDS(cmd *cobra.Command, uri *url.URL) (string, error) {
-	var signers []ssh.Signer
+func GetUDS(cmd *cobra.Command, uri *url.URL, iden string) (string, error) {
+	cfg, err := Validate(uri, iden)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to validate")
+	}
+	dial, err := ssh.Dial("tcp", uri.Host, cfg)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to connect")
+	}
+	defer dial.Close()
 
+	session, err := dial.NewSession()
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create new ssh session on %q", uri.Host)
+	}
+	defer session.Close()
+
+	// Override podman binary for testing etc
+	podman := "podman"
+	if v, found := os.LookupEnv("PODMAN_BINARY"); found {
+		podman = v
+	}
+	run := podman + " info --format=json"
+
+	var buffer bytes.Buffer
+	session.Stdout = &buffer
+	if err := session.Run(run); err != nil {
+		return "", err
+	}
+
+	var info define.Info
+	if err := json.Unmarshal(buffer.Bytes(), &info); err != nil {
+		return "", errors.Wrapf(err, "failed to parse 'podman info' results")
+	}
+
+	if info.Host.RemoteSocket == nil || len(info.Host.RemoteSocket.Path) == 0 {
+		return "", errors.Errorf("remote podman %q failed to report its UDS socket", uri.Host)
+	}
+	return info.Host.RemoteSocket.Path, nil
+}
+
+func Validate(uri *url.URL, iden string) (*ssh.ClientConfig, error) {
+	var signers []ssh.Signer
 	passwd, passwdSet := uri.User.Password()
-	if cmd.Flags().Changed("identity") {
-		value := cmd.Flag("identity").Value.String()
+	if iden != "" && iden != " " {
```

Comment:
What does the `iden == " "` value mean?!

At the very least this needs to be very carefully and prominently documented everywhere it appears. But almost certainly it shouldn’t exist; define a new variable/parameter, with an accurate name (a boolean or an enum) that carries the value explicitly and is visible to IDE tools for looking up definitions/uses, instead of hiding it in an untyped string.

Reply:
iden is the path to the ssh key. I can change this to len(iden) < 2. The iden and url arrays have to be the same length to ensure we don't mismatch keys or go out of bound if identity keys are not required on the ssh host.

Reply:
That’s not quite an answer to what I was asking; it’s an answer to _why_ some code exists but not _what_ the code is/does/means.

(It’s important for documentation to say _what_, because it allows abstraction — other programmers can use your code without understand how it works. Usually, the why should easily follow; but if it is not obvious, it’s extremely valuable to document “why” as well, e.g. if it is a workaround.)

Right now, 
- `if iden != "" && iden != " "`, `iden` is “the path to the ssh key” (…to the public key used to validate the server? … to the private key used to authenticate the client? … to the public key corresponding to a private key used to authenticate the client? These are vastly different, and the differences matter. [It doesn’t help that the current code seems to misname the value somewhere in the path stack, but still…])
- `iden == ""` means: [explain this]
- `iden == " "` means: [explain this and how it is different from `""`]

(AFAICS nothing _differentiates_ between `iden == ""` and `iden == " "`, so the code could just use `""` values everywhere without introducing the `" "` case, couldn’t it?

To be explicit, just to make sure there isn’t a disagreement about this, it is possible to construct a 1-element array that contains an `""` element. A lot of Go code treats `""` as “not present” but that’s mostly because Go is a dumb language without optionals and expressing “not present” another way is inconvenient, not because they are truly the same thing.)

---

Changing this to `len(iden) < 2` would be much worse — why couldn’t `iden` legitimately be a single-character path like `x`?

Reply:
I see what you are saying, I will be more explicit. I have been testing using public keys only so that is why my brain just defaulted to that. iden will either be blank (no validation) or an entry. For some reason I made it so when no applicable iden was inserted as a " " versus a "" which was a bad decision on my end. I will documents what iden is and switch this to just say "" versus `"" && " "`

Reply:
the code works with an array of ["'] I just tested it with a new connection and with no connection specified so we should be good, repushing now.

Reply:
(To be fair, I think a fair bit of Podman has not that stringent standards on documentation; it wouldn’t be reasonable to ask you to add documentation for every bit of Podman your PR touches.)

Reply:
@mtrmac of course, I will add as much documentation as possible thank you for pointing all of this out